### PR TITLE
Adding a distributed object abstraction

### PR DESCRIPTION
**Pull request by weilewei**
This PR aims at adding dist_object for HPX.

A distributed object is a single logical object partitioned across a set of localities. With this dist_object functionality, it provides a nice interface for programmers to write a distributed object. The users do not need to write much code in order to build a distributed component in HPX, and the code will be more readable and portable.

This PR contains dist_object itself, comments for Doxygen, many test cases, and one example (matrix transpose). Please review code when you have time, and I will revise accordingly. Thanks a lot!

### CODE REVIEW

File: hpx/lcos/dist_object.hpp [hunk 1]
```diff
@@ -0,0 +1,566 @@
...
+        /// \param construction_type The construction_type in the template parameters
+        /// accepts either Meta_Object, and it is set to All_to_All by defalut
+        /// The Meta_Object option provides meta object registration in the root
+        /// locality and meta object is essentailly a table that can find the
+        /// instances of dist_object in all localities. The All_to_All option only
+        /// locally holds the client and server of the dist_object.
+        /// \param base_name The name of the dist_object, which should be a unique
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type data)
```
Comment (biddisco) [inline hpx/lcos/dist_object.hpp]: There should also be a constructor that takes a vector of localities so that an object can be created on a subset of all

File: hpx/lcos/dist_object.hpp [hunk 2]
```diff
@@ -0,0 +1,566 @@
...
+            std::vector<hpx::id_type> empty;
+            empty.resize(0);
+            return empty;
+        }
+
+        std::vector<hpx::id_type> registration(
+            std::size_t source_loc, hpx::id_type id)
+        {
+            {
+                // TODO: duplicated lock: the namespace has lock guard already
```
Comment (biddisco) [inline hpx/lcos/dist_object.hpp]: What is the meaning of this comment? Obsolete?

File: hpx/include/dist_object.hpp [hunk 3]
```diff
@@ -0,0 +1,11 @@
+//  Copyright (c) 2019 Weile Wei
+//
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#if !defined(HPX_DIST_OBJECT_APRIl_12_2019_0412PM)
+#define HPX_DIST_OBJECT_APRIl_12_2019_0412PM
+
+#include <hpx/lcos/dist_object.hpp>
```
Comment (biddisco) [inline hpx/include/dist_object.hpp]: The name `dist_object` doesn't seem right. I would prefer `distributed_object` - we don't use abbreviations in other classes generally.

File: tests/unit/lcos/barrier.cpp [hunk 4]
```diff
@@ -12,12 +12,15 @@
...
 #include <cstddef>
+#include <cstdint>
 #include <functional>
 #include <string>
+#include <utility>
 #include <vector>
 
 ///////////////////////////////////////////////////////////////////////////////
-void barrier_test(std::size_t num, std::size_t rank, std::atomic<std::size_t>& c)
+void barrier_test(
```
Comment (biddisco) [inline tests/unit/lcos/barrier.cpp]: Are changes to barrier related to the PR? I suspect a bit of reformatting (whitespace) has crept in. Since you're adding something new, it might be better to remove unrelated changes from the patch.

File: hpx/lcos/dist_object.hpp [hunk 5]
```diff
@@ -0,0 +1,566 @@
...
+}}}
+
+#define REGISTER_DIST_OBJECT_PART_DECLARATION(type)                            \
+    HPX_REGISTER_ACTION_DECLARATION(                                           \
+        hpx::lcos::server::dist_object_part<type>::fetch_action,               \
+        HPX_PP_CAT(__dist_object_part_fetch_action_, type));
+
+/**/
+
+#define REGISTER_DIST_OBJECT_PART(type)                                        \
```
Comment (biddisco) [inline hpx/lcos/dist_object.hpp]: Also naming here, I'd prefer (even though it's a bit long winded)
`REGISTER_DISTRIBUTED_OBJECT_PART` ?
and so on for other macros

**Comment by biddisco** [comment]
Also - general question - can any of the partitioned vector + friends, be integrated with this code so as to reduce any duplicated functionality?

File: examples/transpose/transpose_dist_object.cpp [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+///////////////////////////////////////////////////////////////////////////
+/// This is an example of distributed matrix transposition using dist_object.
+///
+/// A distributed object is a single logical object partitioned across
+/// a set of localities. (A locality is a single node in a cluster or a
+/// NUMA domian in a SMP machine.) Each locality constructs an instance of
```
Comment (msimberg) [inline examples/transpose/transpose_dist_object.cpp]: Tyop (`domian`).

File: examples/transpose/transpose_dist_object.cpp [hunk 7]
```diff
@@ -0,0 +1,380 @@
...
+/// dist_object<T>, where a value of type T represents the value of this
+/// this locality's instance value. Once dist_object<T> is conctructed, it
```
Comment (msimberg) [inline examples/transpose/transpose_dist_object.cpp]: `concructed`

File: examples/transpose/transpose_dist_object.cpp [hunk 8]
```diff
@@ -0,0 +1,380 @@
...
+
+    std::uint64_t order = vm["matrix_size"].as<std::uint64_t>();
+    std::uint64_t iterations = vm["iterations"].as<std::uint64_t>();
+    std::uint64_t num_local_blocks = vm["num_blocks"].as<std::uint64_t>();
+    std::uint64_t tile_size = order;
+
+    if (vm.count("tile_size"))
+        tile_size = vm["tile_size"].as<std::uint64_t>();
+
+    verbose = vm.count("verbose") ? true : false;
```
Comment (msimberg) [inline examples/transpose/transpose_dist_object.cpp]: `<boolean expression> ? true : false` is always redundant ;) You might want to drop everything after `?` or do an explicit `vm.count("verbose") > 0`.

**Comment by msimberg** [comment]
Could you please add (documented) header files to [this list](https://github.com/STEllAR-GROUP/hpx/blob/600f3211fb9ae32a58ed919ca51fce004d310c85/docs/CMakeLists.txt#L7), please?

File: examples/transpose/transpose_dist_object.cpp (hunk 8 above)
Comment (hkaiser) [reply examples/transpose/transpose_dist_object.cpp]: @msimberg just writing `verbose = vm.count("verbose");` generates a stupid MSVC warning, so doing an explicit `vm.count("verbose") > 0` seems to be best.

**Comment by hkaiser** [comment]
@weilewei thanks for working on this! I believe the `dist_object` will become a very useful dictionary type for HPX.

@biddisco This is another good candidate (next to `channel`) to expose RDMA capabilities directly to the user.

**Comment by hkaiser** [comment]
> Also - general question - can any of the partitioned vector + friends, be integrated with this code so as to reduce any duplicated functionality?

@biddisco We could potentially re-implement `partitioned_vector` on top of the distributed object. That might be  even desirable once `dist_object` has been made RDMA aware...

**Comment by weilewei** [comment]
> Also - general question - can any of the partitioned vector + friends, be integrated with this code so as to reduce any duplicated functionality?

Hi John, thanks for all your detailed comments above and I am fixing on every single issue mentioned here. However, I don't think I understand this quoted question well, could you please explain more and provide some small example that can help me understand better, if possible?

What does a `partitioned vector` mean? Do you mean a distributed vector that each locality owns part of the data of the vector? Also, what does the `friend` mean exactly? Also, can you point out where does the duplicated code exist or potentially will be? Any suggestion will help for my next fix. Thanks!

File: hpx/lcos/dist_object.hpp [hunk 9]
```diff
@@ -0,0 +1,566 @@
...
+        /// string across the localities
+        /// \param data The data of the type T of the dist_object
+        dist_object(std::string base, data_type&& data)
+          : base_type(create_server(std::move(data)))
+          , base_(base)
+        {
+            basename_registration_helper(base);
+        }
+        /// \cond NOINTERNAL
+        dist_object(hpx::future<hpx::id_type>&& id)
```
Comment (sithhell) [inline hpx/lcos/dist_object.hpp]: what's the semantic of this constructor?

File: hpx/lcos/dist_object.hpp [hunk 10]
```diff
@@ -0,0 +1,566 @@
...
+    template <typename T>
+    class dist_object_part<T&>
+      : public hpx::components::locking_hook<
+            hpx::components::component_base<dist_object_part<T&>>>
+    {
+    public:
+        typedef T& data_type;
+        dist_object_part() {}
+
+        dist_object_part(data_type data)
```
[the remark this answers is not in the document]
Comment (sithhell) [reply hpx/lcos/dist_object.hpp]: Ok, after finally understanding the purpose ... it makes more sense... is there any usecase?

File: hpx/lcos/dist_object.hpp [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+    /// The dist_object is a single logical object partitioned over a set of
+    /// localities/nodes/machines, where every locality shares the same global
+    /// name locality for the distributed object (i.e. a universal name), but
+    /// owns its local value. In other words, local data of the distributed
+    /// object can be different, but they share access to one another's data
+    /// globally.
+    template <typename T, construction_type C = construction_type::All_to_All>
+    class dist_object
+      : hpx::components::client_base<dist_object<T>,
+            server::dist_object_part<T>>
```
Comment (sithhell) [inline hpx/lcos/dist_object.hpp]: With being a client_base, we also get serializability. What does it mean to serialize a `dist_object`?

File: hpx/lcos/dist_object.hpp [hunk 12]
```diff
@@ -0,0 +1,566 @@
...
+            hpx::id_type lookup = get_basename_helper(idx);
+            typedef typename server::dist_object_part<T&>::fetch_ref_action
+                action_type;
+            return hpx::async<action_type>(lookup);
+        }
+        /// \cond NOINTERNAL
+    private:
+        mutable std::shared_ptr<server::dist_object_part<T&>> ptr;
+        std::string base_;
+        void ensure_ptr() const
```
Comment (sithhell) [inline hpx/lcos/dist_object.hpp]: Why have this lazily? An instance to `dist_object` is always local (IIUC), so upon construction, we can have this pointer and save us the condition here.

File: hpx/lcos/dist_object.hpp [hunk 13]
```diff
@@ -0,0 +1,566 @@
...
+            {
+                ptr = hpx::get_ptr<server::dist_object_part<T&>>(
+                    hpx::launch::sync, this->get_id());
+            }
+        }
+
+    private:
+        std::vector<hpx::id_type> basename_list;
+        hpx::id_type get_basename_helper(int idx)
+        {
```
Comment (sithhell) [inline hpx/lcos/dist_object.hpp]: This needs to be protected with a mutex or documented as being unsafe.

**Comment by biddisco** [comment]
> What does a `partitioned vector` mean? Do you mean a distributed vector that each locality owns part of the data of the vector? Also, what does the `friend` mean exactly? Also, can you point out where does the duplicated code exist or potentially will be? Any suggestion will help for my next fix. 

What I mean is that there is a class in HPX called partitioned_vector that is a distributed vector, it probably has a lot of similarities to your class (or parts of it) - here's a quick list of some code that either uses it or depends on it - have a look at these just to see if there's any duplicated functionality.
(By friends, I just mean that there are lot's of related classes/files in HPX that  are connected)

```
git grep -l partitioned_vector | grep -v docs | sort
...
hpx/components/containers/container_distribution_policy.hpp
hpx/components/containers/partitioned_vector/detail/view_element.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_component_decl.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_component.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_component_impl.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_decl.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_fwd.hpp
hpx/components/containers/partitioned_vector/partitioned_vector.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_impl.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_local_view.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_local_view_iterator.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_predef.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_view.hpp
hpx/components/containers/partitioned_vector/partitioned_vector_view_iterator.hpp
hpx/components/containers/unordered/unordered_map.hpp
hpx/components/containers/unordered/unordered_map_segmented_iterator.hpp
hpx/include/partitioned_vector.hpp
hpx/include/partitioned_vector_predef.hpp
hpx/include/partitioned_vector_view.hpp
hpx/include/serialization.hpp
hpx/runtime/serialization/partitioned_vector.hpp
src/components/containers/CMakeLists.txt
src/components/containers/partitioned_vector/CMakeLists.txt
src/components/containers/partitioned_vector/partitioned_vector_component.cpp
src/components/containers/partitioned_vector/partitioned_vector_component_double.cpp
src/components/containers/partitioned_vector/partitioned_vector_component_int.cpp
src/components/containers/partitioned_vector/partitioned_vector_component_std_string.cpp
tests/performance/local/CMakeLists.txt
tests/performance/local/partitioned_vector_foreach.cpp
tests/performance/network/algorithms/minmax_element_performance.cpp
tests/regressions/components/CMakeLists.txt
tests/regressions/components/partitioned_vector_2201.cpp
tests/unit/component/CMakeLists.txt
tests/unit/component/coarray_all_reduce.cpp
tests/unit/component/coarray.cpp
tests/unit/component/partitioned_vector_subview.cpp
tests/unit/component/partitioned_vector_view.cpp
tests/unit/component/partitioned_vector_view_iterator.cpp
tests/unit/parallel/segmented_algorithms/CMakeLists.txt
tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_difference1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_difference2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_find1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_find2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_all_of1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_all_of2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_any_of1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_any_of2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_copy.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_exclusive_scan2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_exclusive_scan.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_fill.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_find2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_find.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_for_each.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_handle_values.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_inclusive_scan2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_inclusive_scan.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_iter.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_move.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_none1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_none2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_reduce.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_scan.hpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_target.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_binary1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_binary2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce_binary1.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce_binary2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_scan2.cpp
tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_scan.cpp
tests/unit/parallel/segmented_algorithms/test_transform_binary2.hpp
tests/unit/parallel/segmented_algorithms/test_transform_binary.hpp
```

File: hpx/lcos/dist_object.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (hkaiser) [reply hpx/lcos/dist_object.hpp]: @sithhell In Phylanx we hand around Blaze matrices during evaluation. Now, if those matrices are actually just tiles of a bigger matrix and the execution-tree is duplicated across the localities (evaluated SPMD style), then we can simply use the `dist_object<T&>` to attach it to the existing Blaze matrix tile on each of the localities and perform global operations.

File: hpx/lcos/dist_object.hpp (hunk 1 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: Now it can accept a list of sub-localities.
```cpp
        distributed_object(std::string base, data_type const& data, std::vector<size_t> sub_localities = all_localities())
```
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L301

File: hpx/lcos/dist_object.hpp (hunk 2 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: Thanks for attention! Removed now.

File: hpx/include/dist_object.hpp (hunk 3 above)
Comment (weilewei) [reply hpx/include/dist_object.hpp]: I changed all the naming from `dist_object` to `distributed_object`.
https://github.com/STEllAR-GROUP/hpx/blob/dist_object_pr/hpx/lcos/distributed_object.hpp#L301
Please review

File: tests/unit/lcos/barrier.cpp (hunk 4 above)
Comment (weilewei) [reply tests/unit/lcos/barrier.cpp]: Sorry, by mistake. I removed.

File: hpx/lcos/dist_object.hpp (hunk 5 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: Updated as well. Everything is now named `distributed_object`, including file name.

**Comment by weilewei** [comment]
> > What does a `partitioned vector` mean? Do you mean a distributed vector that each locality owns part of the data of the vector? Also, what does the `friend` mean exactly? Also, can you point out where does the duplicated code exist or potentially will be? Any suggestion will help for my next fix.
> 
> What I mean is that there is a class in HPX called partitioned_vector that is a distributed vector, it probably has a lot of similarities to your class (or parts of it) - here's a quick list of some code that either uses it or depends on it - have a look at these just to see if there's any duplicated functionality.
> (By friends, I just mean that there are lot's of related classes/files in HPX that are connected)
> 
> ```
> git grep -l partitioned_vector | grep -v docs | sort
> ...
> hpx/components/containers/container_distribution_policy.hpp
> hpx/components/containers/partitioned_vector/detail/view_element.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_component_decl.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_component.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_component_impl.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_decl.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_fwd.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_impl.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_local_view.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_local_view_iterator.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_predef.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_segmented_iterator.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_view.hpp
> hpx/components/containers/partitioned_vector/partitioned_vector_view_iterator.hpp
> hpx/components/containers/unordered/unordered_map.hpp
> hpx/components/containers/unordered/unordered_map_segmented_iterator.hpp
> hpx/include/partitioned_vector.hpp
> hpx/include/partitioned_vector_predef.hpp
> hpx/include/partitioned_vector_view.hpp
> hpx/include/serialization.hpp
> hpx/runtime/serialization/partitioned_vector.hpp
> src/components/containers/CMakeLists.txt
> src/components/containers/partitioned_vector/CMakeLists.txt
> src/components/containers/partitioned_vector/partitioned_vector_component.cpp
> src/components/containers/partitioned_vector/partitioned_vector_component_double.cpp
> src/components/containers/partitioned_vector/partitioned_vector_component_int.cpp
> src/components/containers/partitioned_vector/partitioned_vector_component_std_string.cpp
> tests/performance/local/CMakeLists.txt
> tests/performance/local/partitioned_vector_foreach.cpp
> tests/performance/network/algorithms/minmax_element_performance.cpp
> tests/regressions/components/CMakeLists.txt
> tests/regressions/components/partitioned_vector_2201.cpp
> tests/unit/component/CMakeLists.txt
> tests/unit/component/coarray_all_reduce.cpp
> tests/unit/component/coarray.cpp
> tests/unit/component/partitioned_vector_subview.cpp
> tests/unit/component/partitioned_vector_view.cpp
> tests/unit/component/partitioned_vector_view_iterator.cpp
> tests/unit/parallel/segmented_algorithms/CMakeLists.txt
> tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_difference1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_difference2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_find1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_adjacent_find2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_all_of1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_all_of2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_any_of1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_any_of2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_copy.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_exclusive_scan2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_exclusive_scan.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_fill.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_find2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_find.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_for_each.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_handle_values.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_inclusive_scan2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_inclusive_scan.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_iter.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_move.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_none1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_none2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_reduce.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_scan.hpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_target.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_binary1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_binary2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce_binary1.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_reduce_binary2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_scan2.cpp
> tests/unit/parallel/segmented_algorithms/partitioned_vector_transform_scan.cpp
> tests/unit/parallel/segmented_algorithms/test_transform_binary2.hpp
> tests/unit/parallel/segmented_algorithms/test_transform_binary.hpp
> ```

As our conversation with Dr. Kaiser, it seems we will update `partitioned_vector` in the future. Please suggest, if you have any ideas. Thanks!

File: examples/transpose/transpose_dist_object.cpp (hunk 6 above)
Comment (weilewei) [reply examples/transpose/transpose_dist_object.cpp]: Thanks for your attention! I updated.

File: examples/transpose/transpose_dist_object.cpp (hunk 7 above)
Comment (weilewei) [reply examples/transpose/transpose_dist_object.cpp]: I also modified some descriptions for constructors and member functions. Please review.
https://github.com/STEllAR-GROUP/hpx/blob/dist_object_pr/hpx/lcos/distributed_object.hpp#L301

File: examples/transpose/transpose_dist_object.cpp (hunk 8 above)
Comment (weilewei) [reply examples/transpose/transpose_dist_object.cpp]: @msimberg @hkaiser please check if this looks good for you. Thanks!
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/examples/transpose/transpose_distributed_object.cpp#L87

**Comment by weilewei** [comment]
> Could you please add (documented) header files to [this list](https://github.com/STEllAR-GROUP/hpx/blob/600f3211fb9ae32a58ed919ca51fce004d310c85/docs/CMakeLists.txt#L7), please?

added here: https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/docs/CMakeLists.txt#L147

File: hpx/lcos/dist_object.hpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: @sithhell Right. Thanks @hkaiser for your explanation. We need it for Phylan project where requires wraps `distributed_object` to an existing object, like `Blaze` matrix

File: hpx/lcos/dist_object.hpp [hunk 14]
```diff
@@ -0,0 +1,566 @@
...
+      : hpx::components::client_base<meta_object, meta_object_server>
+    {
+    public:
+        typedef hpx::components::client_base<meta_object, meta_object_server>
+            base_type;
+
+        meta_object(std::string basename)
+          : base_type(hpx::local_new<meta_object_server>())
+        {
+            if (hpx::get_locality_id() == 0)
```
[the remark this answers is not in the document]
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: Thanks for your suggestion. Actually, it can be any locality and being in an arbitrary locality adds flexibility. See here: https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L219

Now, we sets the root to the first locality of the given sub-locality list. For example, If all localities (default option) are given to construct `distributed_object`, then the root will be locality 0. Say, if you provide localities `1, 3, 5`, then the locality `1` will be the root for the meta_object. Let me know if this fits for you.

File: hpx/lcos/dist_object.hpp (hunk 9 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: If I remeber correctly from @hkaiser, this line generates boilerplate code for the component client. So no error will occur.

File: hpx/lcos/dist_object.hpp (hunk 11 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: Thanks for the notice. I added code here:
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L460

File: hpx/lcos/dist_object.hpp (hunk 12 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: Good suggestion. I put the `ensure_ptr` inside the constructor now.
https://github.com/STEllAR-GROUP/hpx/blob/b1c8f32ae0627852cf02f96013a70b443f03a066/hpx/lcos/distributed_object.hpp#L310
Let me know if you have any suggestion.

File: hpx/lcos/dist_object.hpp (hunk 13 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: I might miss this suggestion at this point. Do you mean that I need to place a `mutex` on `std::vector<hpx::id_type> basename_list;`. Or how I should write the code to fix it?

File: hpx/lcos/dist_object.hpp (hunk 11 above)
Comment (hkaiser) [reply hpx/lcos/dist_object.hpp]: There is no need to mark this function as `HPX_FORCEINLINE`.

File: hpx/lcos/dist_object.hpp (hunk 12 above)
Comment (hkaiser) [reply hpx/lcos/dist_object.hpp]: Yes, and with #3827 merged this does not incur any undue overhead.

File: hpx/lcos/dist_object.hpp (hunk 11 above)
Comment (weilewei) [reply hpx/lcos/dist_object.hpp]: @hkaiser removed.

File: hpx/lcos/dist_object.hpp (hunk 9 above)
Comment (sithhell) [reply hpx/lcos/dist_object.hpp]: well, I think this is a design error. Is a `distributed_object` ever to be constructed from a `hpx::future<hpx::id_type>&& id`, what does this mean? I think this is a similar problem as with the serialize function.

File: hpx/lcos/dist_object.hpp (hunk 13 above)
Comment (sithhell) [reply hpx/lcos/dist_object.hpp]: Yes, you need to protect the member to be modified and read concurrently.

File: hpx/lcos/dist_object.hpp (hunk 9 above)
Comment (hkaiser) [reply hpx/lcos/dist_object.hpp]: This might be needed for `distributed_object<>::base_type::then` to compile. We could make the base class a `friend` and hide the constructor as being `private`.

File: hpx/lcos/dist_object.hpp (hunk 13 above)
Comment (hkaiser) [reply hpx/lcos/dist_object.hpp]: We should also turn this into a lazy operation, I think.

File: hpx/lcos/dist_object.hpp (hunk 9 above)
Comment (sithhell) [reply hpx/lcos/dist_object.hpp]: Does `distributed_object<>::base_type::then` make any sense? The construction of a `distributed_object` should be almost immediate, no?
I think there are more functions from `hpx::components::client_base` that don't make sense for `distributed_object`: All assignment operators taking a `hpx::id_type` or `hpx::future<hpx::id_type>`, `create` and `create_colocated`, `register_as`, `connect_to`, `reset_registered_name`. I think `hpx::components::client_base` is the wrong abstraction for `distributed_object`.

Comment (hkaiser) [reply hpx/lcos/dist_object.hpp]: `register_as` is definitely needed as siblings on other localities connect through the symbol namespace if needed. But I concur to your point, `client_base` is too heavy for this specific use case.

**Comment by hkaiser** [comment]
We have added a simplified version of this to Phylanx (see https://github.com/STEllAR-GROUP/phylanx/pull/990) for now. Once we're happy with what we have over there we can reconsider adding it to HPX proper. Closing this for now.
